File: pluck/pipelines/convert.py
"""Format convert: remux (stream-copy) when possible, else transcode to convert_to."""
from pathlib import Path

from ..models import CONVERT_TARGETS
from .base import JobCtx, download_source, run_ffmpeg, trim_if_requested

AUDIO_TARGETS = {"mp3", "m4a", "opus", "wav", "flac"}

# Transcode recipes per target (audio targets re-encode audio; video targets try copy first).
_AUDIO_CODEC = {
    "mp3": ["-c:a", "libmp3lame", "-b:a", "192k"],
    "m4a": ["-c:a", "aac", "-b:a", "192k"],
    "opus": ["-c:a", "libopus", "-b:a", "160k"],
    "wav": ["-c:a", "pcm_s16le"],
    "flac": ["-c:a", "flac"],
}


def convert_file(src: Path, target: str) -> Path:
    target = (target or "").lower().lstrip(".")
    if target not in CONVERT_TARGETS:
        raise RuntimeError(f"unsupported convert target: {target!r}")
    out = src.with_name(src.stem + "-conv." + target)

    if target in AUDIO_TARGETS:
        args = ["-i", str(src), "-vn", *_AUDIO_CODEC[target], str(out)]
        r = run_ffmpeg(args, timeout=600)
    else:
        # video container: try stream copy first (fast, lossless), fall back to transcode
        r = run_ffmpeg(["-i", str(src), "-c", "copy", str(out)], timeout=600)
        if r.returncode != 0 or not out.exists() or out.stat().st_size == 0:
            vcodec = ["-c:v", "libvpx-vp9", "-c:a", "libopus"] if target == "webm" else ["-c:v", "libx264", "-c:a", "aac"]
            r = run_ffmpeg(["-i", str(src), *vcodec, str(out)], timeout=1200)

    if r.returncode == 0 and out.exists() and out.stat().st_size > 0:
        src.unlink(missing_ok=True)
        return out
    raise RuntimeError("convert failed: " + r.stderr.decode("utf-8", "ignore")[-200:])


def run(ctx: JobCtx) -> Path:
    req = ctx.req
    target = (req.convert_to or "").lower().lstrip(".")
    if target not in CONVERT_TARGETS:
        raise RuntimeError(f"unsupported convert target: {req.convert_to!r}")
    src = download_source(ctx, audio_only=target in AUDIO_TARGETS)
    ctx.update(status="processing")
    src = trim_if_requested(src, req)
    return convert_file(src, target)

File: node/src/pipelines/convert.js
/** Format convert: remux (stream-copy) when possible, else transcode to convert_to.
 * Mirrors pluck/pipelines/convert.py. */
import path from "node:path";
import { statSync } from "node:fs";

import { CONVERT_TARGETS } from "../models.js";
import { downloadSource, runFfmpeg, trimIfRequested, unlinkQuiet } from "./base.js";

export const AUDIO_TARGETS = new Set(["mp3", "m4a", "opus", "wav", "flac"]);

// Transcode recipes per target (audio targets re-encode audio; video targets try copy first).
const AUDIO_CODEC = {
  mp3: ["-c:a", "libmp3lame", "-b:a", "192k"],
  m4a: ["-c:a", "aac", "-b:a", "192k"],
  opus: ["-c:a", "libopus", "-b:a", "160k"],
  wav: ["-c:a", "pcm_s16le"],
  flac: ["-c:a", "flac"],
};

function ok(p) {
  try { return statSync(p).size > 0; } catch { return false; }
}

export async function convertFile(src, target) {
  target = (target || "").toLowerCase().replace(/^\.+/, "");
  if (!CONVERT_TARGETS.has(target)) {
    throw new Error(`unsupported convert target: '${target}'`);
  }
  const dir = path.dirname(src);
  const stem = path.basename(src, path.extname(src));
  const out = path.join(dir, `${stem}-conv.${target}`);

  let r;
  if (AUDIO_TARGETS.has(target)) {
    r = await runFfmpeg(["-i", src, "-vn", ...AUDIO_CODEC[target], out], 600_000);
  } else {
    // video container: try stream copy first (fast, lossless), fall back to transcode
    r = await runFfmpeg(["-i", src, "-c", "copy", out], 600_000);
    if (r.returncode !== 0 || !ok(out)) {
      const vcodec = target === "webm"
        ? ["-c:v", "libvpx-vp9", "-c:a", "libopus"]
        : ["-c:v", "libx264", "-c:a", "aac"];
      r = await runFfmpeg(["-i", src, ...vcodec, out], 1_200_000);
    }
  }

  if (r.returncode === 0 && ok(out)) {
    unlinkQuiet(src);
    return out;
  }
  throw new Error("convert failed: " + r.stderr.slice(-200));
}

export async function run(ctx) {
  const req = ctx.req;
  const target = (req.convert_to || "").toLowerCase().replace(/^\.+/, "");
  if (!CONVERT_TARGETS.has(target)) {
    throw new Error(`unsupported convert target: '${req.convert_to}'`);
  }
  let src = await downloadSource(ctx, { audioOnly: AUDIO_TARGETS.has(target) });
  ctx.update({ status: "processing" });
  src = await trimIfRequested(src, req);
  return convertFile(src, target);
}
